File: clients/python/clausters/base/absobject.py
"""Operator-overloading base (port of ``sc3/base/absobject.py``).

``AbstractObject`` overloads Python's arithmetic, comparison and bitwise
operators and exposes named math methods, routing every one through four
dispatch hooks a subclass implements:

- ``_compose_unop(selector)``           — unary
- ``_compose_binop(selector, other)``   — binary
- ``_rcompose_binop(selector, other)``  — reflected binary (``other <op> self``)
- ``_compose_narop(selector, *args)``   — n-ary

The ``selector`` strings are the same operator names used by
`clausters.base.builtins` (value side) and by
`clausters.defs.signals` (graph side) — so the *same* expression composes
either concrete numbers or a Faust/UGen graph depending on the subclass. This
is what lets one piece of code describe both.
"""


class AbstractObject:
    # --- dispatch hooks (subclass responsibility) ---

    def _compose_unop(self, selector):
        raise NotImplementedError(f"{type(self).__name__}._compose_unop")

    def _compose_binop(self, selector, other):
        raise NotImplementedError(f"{type(self).__name__}._compose_binop")

    def _rcompose_binop(self, selector, other):
        raise NotImplementedError(f"{type(self).__name__}._rcompose_binop")

    def _compose_narop(self, selector, *args):
        raise NotImplementedError(f"{type(self).__name__}._compose_narop")

    # --- unary operators ---

    def __neg__(self): return self._compose_unop("neg")
    def __pos__(self): return self
    def __abs__(self): return self._compose_unop("abs")
    def __floor__(self): return self._compose_unop("floor")
    def __ceil__(self): return self._compose_unop("ceil")

    # --- binary operators ---

    def __add__(self, other): return self._compose_binop("add", other)
    def __radd__(self, other): return self._rcompose_binop("add", other)
    def __sub__(self, other): return self._compose_binop("sub", other)
    def __rsub__(self, other): return self._rcompose_binop("sub", other)
    def __mul__(self, other): return self._compose_binop("mul", other)
    def __rmul__(self, other): return self._rcompose_binop("mul", other)
    def __truediv__(self, other): return self._compose_binop("div", other)
    def __rtruediv__(self, other): return self._rcompose_binop("div", other)
    def __mod__(self, other): return self._compose_binop("mod", other)
    def __rmod__(self, other): return self._rcompose_binop("mod", other)
    def __pow__(self, other): return self._compose_binop("pow", other)
    def __rpow__(self, other): return self._rcompose_binop("pow", other)
    def __lshift__(self, other): return self._compose_binop("lshift", other)
    def __rlshift__(self, other): return self._rcompose_binop("lshift", other)
    def __rshift__(self, other): return self._compose_binop("rshift", other)
    def __rrshift__(self, other): return self._rcompose_binop("rshift", other)
    def __and__(self, other): return self._compose_binop("bitand", other)
    def __rand__(self, other): return self._rcompose_binop("bitand", other)
    def __or__(self, other): return self._compose_binop("bitor", other)
    def __ror__(self, other): return self._rcompose_binop("bitor", other)
    def __xor__(self, other): return self._compose_binop("bitxor", other)
    def __rxor__(self, other): return self._rcompose_binop("bitxor", other)

    # --- comparison operators (return composed objects, not bools) ---

    def __lt__(self, other): return self._compose_binop("lt", other)
    def __le__(self, other): return self._compose_binop("le", other)
    def __gt__(self, other): return self._compose_binop("gt", other)
    def __ge__(self, other): return self._compose_binop("ge", other)

    # --- named unary methods ---

    def abs(self): return self._compose_unop("abs")
    def neg(self): return self._compose_unop("neg")
    def floor(self): return self._compose_unop("floor")
    def ceil(self): return self._compose_unop("ceil")
    def sin(self): return self._compose_unop("sin")
    def cos(self): return self._compose_unop("cos")
    def tan(self): return self._compose_unop("tan")
    def asin(self): return self._compose_unop("asin")
    def acos(self): return self._compose_unop("acos")
    def atan(self): return self._compose_unop("atan")
    def exp(self): return self._compose_unop("exp")
    def log(self): return self._compose_unop("log")
    def log10(self): return self._compose_unop("log10")
    def sqrt(self): return self._compose_unop("sqrt")
    def asinteger(self): return self._compose_unop("as_int")
    def asfloat(self): return self._compose_unop("as_float")
    def log2(self): return self._compose_unop("log2")
    def squared(self): return self._compose_unop("squared")
    def cubed(self): return self._compose_unop("cubed")
    def reciprocal(self): return self._compose_unop("recip")
    def frac(self): return self._compose_unop("frac")
    def sign(self): return self._compose_unop("sign")
    def sinh(self): return self._compose_unop("sinh")
    def cosh(self): return self._compose_unop("cosh")
    def tanh(self): return self._compose_unop("tanh")
    def distort(self): return self._compose_unop("distort")
    def softclip(self): return self._compose_unop("softclip")
    def midicps(self): return self._compose_unop("midicps")
    def cpsmidi(self): return self._compose_unop("cpsmidi")
    def midiratio(self): return self._compose_unop("midiratio")
    def ratiomidi(self): return self._compose_unop("ratiomidi")
    def dbamp(self): return self._compose_unop("dbamp")
    def ampdb(self): return self._compose_unop("ampdb")
    def octcps(self): return self._compose_unop("octcps")
    def cpsoct(self): return self._compose_unop("cpsoct")

    # --- named binary methods ---

    def min(self, other): return self._compose_binop("min", other)
    def max(self, other): return self._compose_binop("max", other)
    def atan2(self, other): return self._compose_binop("atan2", other)
    def pow(self, other): return self._compose_binop("pow", other)
    def mod(self, other): return self._compose_binop("mod", other)
    def hypot(self, other): return self._compose_binop("hypot", other)
    def hypotapx(self, other): return self._compose_binop("hypot_apx", other)
    def clip2(self, other): return self._compose_binop("clip2", other)
    def fold2(self, other): return self._compose_binop("fold2", other)
    def wrap2(self, other): return self._compose_binop("wrap2", other)
    def gcd(self, other): return self._compose_binop("gcd", other)
    def lcm(self, other): return self._compose_binop("lcm", other)
    def excess(self, other): return self._compose_binop("excess", other)
    def absdif(self, other): return self._compose_binop("absdif", other)
    def thresh(self, other): return self._compose_binop("thresh", other)
    def round(self, other): return self._compose_binop("round", other)
    def trunc(self, other): return self._compose_binop("trunc", other)
    def ring1(self, other): return self._compose_binop("ring1", other)
    def ring2(self, other): return self._compose_binop("ring2", other)
    def ring3(self, other): return self._compose_binop("ring3", other)
    def ring4(self, other): return self._compose_binop("ring4", other)
    def sumsqr(self, other): return self._compose_binop("sumsqr", other)
    def difsqr(self, other): return self._compose_binop("difsqr", other)
    def sqrsum(self, other): return self._compose_binop("sqrsum", other)
    def sqrdif(self, other): return self._compose_binop("sqrdif", other)

File: clients/web/src/base/absobject.ts
// The operator surface every expression composes with (mirrors
// `clausters/base/absobject.py`).
//
// **Composition is by method, not by operator.** TypeScript has no operator
// overloading, so where the Python client writes `sine(freq) * amp` this one
// writes `sine(freq).mul(amp)`, and every other operator or math method
// (`mod`, `min`/`max`, comparisons, `.sin()`, `.midicps()`, `.distort()` …) is
// a method carrying the same operator **name** the wire uses — so the two
// clients emit identical specs.
//
// Every method routes through two hooks a subclass implements, `binop` and
// `unop`, which is what lets the *same* written expression compose a UGen
// graph (`defs/ugens/graph`) or a per-bin program (`defs/pv_expr`) depending
// on the subclass. The selectors are the shared builtins table's names, so
// this class is about the **vocabulary** and knows nothing about UGens.
//
// It lives here, beside the value functions that compute the same operators
// (`base/builtins`), for the reason the Python client keeps `AbstractObject`
// in `base/`: it is neither a def nor a graph. The two type parameters after
// `TSelf` is what a typed language needs and an untyped one does not: [`Fan`]
// says which operand *fans a result out* (a channel list, in the UGen graph)
// and into what, and is `Fan<never, never>` where nothing does.

/**
 * Which operand kind fans a result out, and into what. One type because the
 * two only ever travel together: `Fan<ChannelList, ChannelList>` for the UGen
 * graph, the default for an expression where a method always answers itself.
 */
export interface Fan<TList, TFan> {
    list: TList;
    fan: TFan;
}

/**
 * A math method's result: an operand of the fanning kind yields `F["fan"]`,
 * anything else keeps the receiver's shape.
 */
export type Composed<TSelf, TOther, F extends Fan<unknown, unknown>> =
    TOther extends F["list"] ? F["fan"] : TSelf;

/**
 * The operator vocabulary as methods, over the two composition hooks.
 *
 *  * `TSelf` is what a unary op answers, `TOperand` what a binary one accepts,
 * and `F` the fan-out pair described above.
 */
export abstract class AbstractObject<
    TSelf,
    TOperand,
    F extends Fan<unknown, unknown> = Fan<never, never>,
> {
    protected abstract binop<T extends TOperand>(
        selector: string,
        other: T,
    ): Composed<TSelf, T, F>;
    protected abstract unop(selector: string): TSelf;

    // --- binary ---
    add<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("add", x); }
    sub<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("sub", x); }
    mul<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("mul", x); }
    div<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("div", x); }
    mod<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("mod", x); }
    pow<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("pow", x); }
    min<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("min", x); }
    max<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("max", x); }
    atan2<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("atan2", x); }
    gt<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("gt", x); }
    lt<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("lt", x); }
    ge<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("ge", x); }
    le<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("le", x); }
    eq<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("eq", x); }
    ne<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("ne", x); }
    bitand<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("bitand", x); }
    bitor<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("bitor", x); }
    bitxor<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("bitxor", x); }
    leftshift<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("lshift", x); }
    rightshift<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("rshift", x); }
    hypot<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("hypot", x); }
    /** The cheap hypotenuse approximation (`hypotapx` on the wire). */
    hypotapx<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("hypotapx", x); }
    ring1<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("ring1", x); }
    ring2<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("ring2", x); }
    ring3<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("ring3", x); }
    ring4<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("ring4", x); }
    sumsqr<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("sumsqr", x); }
    difsqr<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("difsqr", x); }
    sqrsum<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("sqrsum", x); }
    sqrdif<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("sqrdif", x); }
    absdif<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("absdif", x); }
    thresh<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("thresh", x); }
    clip2<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("clip2", x); }
    excess<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("excess", x); }
    round<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("round", x); }
    trunc<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("trunc", x); }
    fold2<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("fold2", x); }
    wrap2<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("wrap2", x); }
    gcd<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("gcd", x); }
    lcm<T extends TOperand>(x: T): Composed<TSelf, T, F> { return this.binop("lcm", x); }

    // --- unary ---
    neg(): TSelf { return this.unop("neg"); }
    abs(): TSelf { return this.unop("abs"); }
    sin(): TSelf { return this.unop("sin"); }
    cos(): TSelf { return this.unop("cos"); }
    tan(): TSelf { return this.unop("tan"); }
    asin(): TSelf { return this.unop("asin"); }
    acos(): TSelf { return this.unop("acos"); }
    atan(): TSelf { return this.unop("atan"); }
    exp(): TSelf { return this.unop("exp"); }
    log(): TSelf { return this.unop("log"); }
    log10(): TSelf { return this.unop("log10"); }
    log2(): TSelf { return this.unop("log2"); }
    sqrt(): TSelf { return this.unop("sqrt"); }
    floor(): TSelf { return this.unop("floor"); }
    ceil(): TSelf { return this.unop("ceil"); }
    rint(): TSelf { return this.unop("rint"); }
    /** Truncate towards zero to an integer value (`asint` on the wire). */
    asinteger(): TSelf { return this.unop("asint"); }
    /** The identity that documents a value as a float (`asfloat`). */
    asfloat(): TSelf { return this.unop("asfloat"); }
    squared(): TSelf { return this.unop("squared"); }
    cubed(): TSelf { return this.unop("cubed"); }
    reciprocal(): TSelf { return this.unop("recip"); }
    frac(): TSelf { return this.unop("frac"); }
    sign(): TSelf { return this.unop("sign"); }
    sinh(): TSelf { return this.unop("sinh"); }
    cosh(): TSelf { return this.unop("cosh"); }
    tanh(): TSelf { return this.unop("tanh"); }
    distort(): TSelf { return this.unop("distort"); }
    softclip(): TSelf { return this.unop("softclip"); }
    midicps(): TSelf { return this.unop("midicps"); }
    cpsmidi(): TSelf { return this.unop("cpsmidi"); }
    midiratio(): TSelf { return this.unop("midiratio"); }
    ratiomidi(): TSelf { return this.unop("ratiomidi"); }
    dbamp(): TSelf { return this.unop("dbamp"); }
    ampdb(): TSelf { return this.unop("ampdb"); }
    octcps(): TSelf { return this.unop("octcps"); }
    cpsoct(): TSelf { return this.unop("cpsoct"); }
}
